File: content/realm/examples/kotlin/shared/src/commonTest/kotlin/com/mongodb/realm/realmkmmapp/SyncedRealmCRUD.kt
package com.mongodb.realm.realmkmmapp

import io.realm.kotlin.Realm
import io.realm.kotlin.ext.query
import io.realm.kotlin.mongodb.App
import io.realm.kotlin.mongodb.Credentials
import io.realm.kotlin.mongodb.exceptions.CompensatingWriteException
import io.realm.kotlin.mongodb.subscriptions
import io.realm.kotlin.mongodb.sync.SyncConfiguration
import io.realm.kotlin.mongodb.sync.SyncSession
import io.realm.kotlin.mongodb.syncSession
import io.realm.kotlin.query.RealmResults
import io.realm.kotlin.types.RealmObject
import io.realm.kotlin.types.annotations.PrimaryKey
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.runBlocking
import org.mongodb.kbson.ObjectId
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.time.Duration.Companion.seconds

// :replace-start: {
//    "terms": {
//       "SyncItem": "Item"
//    }
// }

class SyncedRealmCRUD : RealmTest() {
    // :snippet-start: flexible-sync-crud-model
    class SyncItem : RealmObject {
        @PrimaryKey
        var _id: ObjectId = ObjectId()
        var ownerId: String = ""
        var itemName: String = ""
        var complexity: Int = 0
    }
    // :snippet-end:

    @Test
    fun subscriptionSetupTest() {
        val credentials = Credentials.anonymous(reuseExisting = false)
        runBlocking {
            // :snippet-start: flexible-sync-subscription-setup
            val app = App.create(FLEXIBLE_APP_ID)
            val user = app.login(credentials)
            val flexSyncConfig = SyncConfiguration.Builder(user, setOf(SyncItem::class))
                // Add subscription
                .initialSubscriptions { realm ->
                    add(
                        // Get Items from Atlas that match the Realm Query Language query.
                        // Uses the queryable field `complexity`.
                        // Query matches objects with complexity less than or equal to 4.
                        realm.query<SyncItem>("complexity <= 4"),
                        "simple-items"
                    )
                }
                .build()
            val syncRealm = Realm.open(flexSyncConfig)
            syncRealm.subscriptions.waitForSynchronization()
            Log.v("Successfully opened realm: ${syncRealm.configuration}")
            // :snippet-end:
            val oid = ObjectId()
            // :snippet-start: successful-write
            // Per the Device Sync permissions, users can only read and write data
            // where the `Item.ownerId` property matches their own user ID.
            val userId = user.id
            val newItem = SyncItem().apply {
                _id = oid // :remove:
                ownerId = userId
                itemName = "This item meets sync criteria"
                complexity = 3
            }

            syncRealm.write {
                // `newItem` is successfully written to the realm and synced to Atlas
                // because its data matches the subscription query (complexity <= 4)
                // and its `ownerId` field matches the user ID.
                copyToRealm(newItem)
            }
            // :snippet-end:
            syncRealm.write {
                val addedItem: RealmResults<SyncItem> = query<SyncItem>().find()
                assertEquals(1, addedItem.size)
                assertEquals(oid, addedItem[0]._id)
                delete(addedItem)
            }
            user.delete()
            syncRealm.close()
        }
    }

   // @Test
    fun compensatingWriteTest() {
        val credentials = Credentials.anonymous(reuseExisting = false)

        val channel = Channel<CompensatingWriteException>()
        // :snippet-start: access-compensating-write
        val syncErrorHandler = SyncSession.ErrorHandler { session, error ->
            runBlocking {
                if (error is CompensatingWriteException) {
                    channel.send(error) // :remove:
                    error.writes.forEach { writeInfo ->
                        val errorMessage = """
                            A write was rejected with a compensating write error
                            The write to object type: ${writeInfo.objectType}
                            With primary key of: ${writeInfo.primaryKey}
                            Was rejected because: ${writeInfo.reason}
                            """.trimIndent()
                        Log.e(errorMessage)
                    }
                }
            }
        }
        // :snippet-end:
        runBlocking {
            val app = App.create(FLEXIBLE_APP_ID)
            val user = app.login(credentials)
            val flexSyncConfig = SyncConfiguration.Builder(user, setOf(SyncItem::class))
                .initialSubscriptions { realm -> add(realm.query<SyncItem>("complexity <= 4"), "simple-items") }
                .errorHandler(syncErrorHandler)
                .build()
            val syncRealm = Realm.open(flexSyncConfig)
            syncRealm.subscriptions.waitForSynchronization()
            Log.v("Successfully opened realm: ${syncRealm.configuration}")
            syncRealm.write { copyToRealm(SyncItem().apply {
                    ownerId = user.id
                    itemName = "This is a compensating write error handling test"
                    complexity = 7 })}
            syncRealm.syncSession.uploadAllLocalChanges(30.seconds)
            val exception: CompensatingWriteException = channel.receiveOrFail()
            exception.message?.let { message ->
                assertTrue(
                    message.contains("[Sync][CompensatingWrite(1033)] Client attempted a write that is outside of permissions or query filters; it has been reverted"),
                    "Exception message did not contain the expected substring"
                )
            }
            assertEquals(1, exception.writes.size)
            channel.close()
            user.delete()
            syncRealm.close()
        }
    }

   // @Test
    fun compensatingWriteQueryTest() {
        val credentials = Credentials.anonymous(reuseExisting = false)
        val channel = Channel<CompensatingWriteException>()
        val syncErrorHandler = SyncSession.ErrorHandler { session, error ->
            runBlocking { if (error is CompensatingWriteException) { channel.send(error) } } }
        runBlocking {
            val app = App.create(FLEXIBLE_APP_ID)
            val user = app.login(credentials)
            val flexSyncConfig = SyncConfiguration.Builder(user, setOf(SyncItem::class))
                .initialSubscriptions { realm -> add(realm.query<SyncItem>("complexity <= 4"), "simple-items") }
                .errorHandler(syncErrorHandler)
                .build()
            val syncRealm = Realm.open(flexSyncConfig)
            syncRealm.subscriptions.waitForSynchronization()
            Log.v("Successfully opened realm: ${syncRealm.configuration}")
            // :snippet-start: write-outside-flexible-sync-query
            // The complexity of this item is `7`. This is outside the bounds
            // of the subscription query, which triggers a compensating write.
            val itemTooComplex = SyncItem().apply {
                ownerId = user.id
                itemName = "This item is too complex"
                complexity = 7
            }
            syncRealm.write {
                copyToRealm(itemTooComplex)
            }
            // :snippet-end:
            syncRealm.syncSession.uploadAllLocalChanges(30.seconds)
            val exception: CompensatingWriteException = channel.receiveOrFail()
            exception.message?.let { message ->
                assertTrue(
                    message.contains("[Sync][CompensatingWrite(1033)] Client attempted a write that is outside of permissions or query filters; it has been reverted"),
                    "Exception message did not contain the expected substring"
                )
            }
            assertEquals(1, exception.writes.size)
            channel.close()
            user.delete()
            syncRealm.close()
        }
    }

  //  @Test
    fun compensatingWritePermissionTest() {
        val credentials = Credentials.anonymous(reuseExisting = false)
        val channel = Channel<CompensatingWriteException>()
        val syncErrorHandler = SyncSession.ErrorHandler { session, error ->
            runBlocking { if (error is CompensatingWriteException) { channel.send(error) } } }
        runBlocking {
            val app = App.create(FLEXIBLE_APP_ID)
            val user = app.login(credentials)
            val flexSyncConfig = SyncConfiguration.Builder(user, setOf(SyncItem::class))
                .initialSubscriptions { realm -> add(realm.query<SyncItem>("complexity <= 4"), "simple-items") }
                .errorHandler(syncErrorHandler)
                .build()
            val syncRealm = Realm.open(flexSyncConfig)
            syncRealm.subscriptions.waitForSynchronization()
            Log.v("Successfully opened realm: ${syncRealm.configuration}")
            // :snippet-start: write-outside-permissions
            // The `ownerId` of this item does not match the `user.id` of the logged-in
            // user. The user does not have permissions to make this write, which
            // triggers a compensating write.
            val itemWithWrongOwner = SyncItem().apply {
                ownerId = "not the current user"
                itemName = "A simple item"
                complexity = 1
            }
            syncRealm.write {
                copyToRealm(itemWithWrongOwner)
            }
            // :snippet-end:
            syncRealm.syncSession.uploadAllLocalChanges(30.seconds)
            val exception: CompensatingWriteException = channel.receiveOrFail()
            exception.message?.let { message ->
                assertTrue(
                    message.contains("[Sync][CompensatingWrite(1033)] Client attempted a write that is outside of permissions or query filters; it has been reverted"),
                    "Exception message did not contain the expected substring"
                )
            }
            assertEquals(1, exception.writes.size)
            channel.close()
            syncRealm.close()
        }
    }
}

// :replace-end:

File: content/realm/examples/ios/Examples/SyncedRealmCRUD.swift
// :replace-start: {
//   "terms": {
//     "FlexibleSync_": "",
//     "APPID_FS_OWNER_READ_WRITE": "YOUR_APP_ID_HERE"
//   }
// }
// swiftlint:disable identifier_name

import XCTest
import RealmSwift

let APPID_FS_OWNER_READ_WRITE = "swiftuiflexsyncexample-dtgeq"

// :snippet-start: flexible-sync-crud-model
class FlexibleSync_Item: Object {
    @Persisted(primaryKey: true) var _id: ObjectId
    @Persisted var ownerId: String
    @Persisted var itemName: String
    @Persisted var complexity: Int
}
// :snippet-end:

class SyncedRealmCRUDCase: XCTestCase {

    override func setUp() {
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    @MainActor
    override func tearDown() async throws {
        let app = App(id: APPID_FS_OWNER_READ_WRITE)

        let user = try await app.login(credentials: Credentials.anonymous)
        var flexSyncConfig = user.flexibleSyncConfiguration()
        flexSyncConfig.objectTypes = [FlexibleSync_Item.self]

        let realm = try await Realm(configuration: flexSyncConfig)
        let subscriptions = realm.subscriptions
        try await subscriptions.update {
            subscriptions.removeAll()
        }
        XCTAssertEqual(subscriptions.count, 0)
    }

    @MainActor
    func testSyncedRealmWriteExamples() async {
        // :snippet-start: flexible-sync-subscription-setup
        let app = App(id: APPID_FS_OWNER_READ_WRITE)

        do {
            let user = try await app.login(credentials: Credentials.anonymous)
            do {
                var flexSyncConfig = user.flexibleSyncConfiguration()
                flexSyncConfig.objectTypes = [FlexibleSync_Item.self]
                let realm = try await Realm(configuration: flexSyncConfig)
                let subscriptions = realm.subscriptions
                try await subscriptions.update {
                    subscriptions.append(
                        QuerySubscription<FlexibleSync_Item>(name: "simple-items") {
                         $0.complexity <= 4
                    })
                }
                print("Successfully opened realm: \(realm)")
                await writeToSyncedRealm(realm: realm, user: user) // :remove:
            } catch {
                print("Failed to open realm: \(error.localizedDescription)")
                // handle error
            }
        } catch {
            fatalError("Login failed: \(error.localizedDescription)")
        }
        // :snippet-end:
    }
    
    @MainActor
    func writeToSyncedRealm(realm: Realm, user: User) async {
        // :snippet-start: successful-write
        // This write falls within the subscription query and complies
        // with the Device Sync permissions, so this write should succeed.
        do {
            let learnRealm = FlexibleSync_Item()
            learnRealm.ownerId = user.id
            learnRealm.itemName = "Learn Realm CRUD stuff"
            learnRealm.complexity = 3
            try realm.write {
                realm.add(learnRealm)
            }
            // :remove-start:
            XCTAssert(realm.objects(FlexibleSync_Item.self).count == 1)
            try realm.write {
                realm.deleteAll()
            }
            // :remove-end:
        } catch {
            print("Failed to write to realm: \(error.localizedDescription)")
        }
        // :snippet-end:
        
        // :snippet-start: write-outside-flexible-sync-query
        do {
            let fixTheBug = FlexibleSync_Item()
            fixTheBug.ownerId = user.id
            fixTheBug.itemName = "Fix the bug with the failing method"
            // The complexity of this item is `7`. This is outside the bounds
            // of the subscription query, so this write triggers a compensating write.
            fixTheBug.complexity = 7
            try realm.write {
                realm.add(fixTheBug)
            }
        } catch {
            print("Failed to write to realm: \(error.localizedDescription)")
        }
        // :snippet-end:
        
        // :snippet-start: write-outside-permissions
        do {
            let itemWithWrongOwner = FlexibleSync_Item()
            // The `ownerId` of this item does not match the `user.id` of the logged-in
            // user. The user does not have permissions to make this write, so
            // it triggers a compensating write.
            itemWithWrongOwner.ownerId = "This string does not match the user.id"
            itemWithWrongOwner.itemName = "Write code that generates a permission error"
            itemWithWrongOwner.complexity = 1
            try realm.write {
                realm.add(itemWithWrongOwner)
            }
        } catch {
            print("Failed to write to realm: \(error.localizedDescription)")
        }
        // :snippet-end:
    }
}
// :replace-end:
